addon/mixins/adapter-fetch.ts

```diff
@@ -0,0 +1,187 @@
+}
+
+export default Mixin.create({
+  headers: undefined as undefined | PlainHeaders,
+  /**
+   * @override
+   */
+  ajaxOptions(url: string, type: Method, options: object): FetchOptions {
+    let hash = (options || {}) as AjaxOptions;
+    hash.url = url;
+    hash.type = type;
+
+    // Add headers set on the Adapter
+    let adapterHeaders = get(this, 'headers');
+    if (adapterHeaders) {
+      hash.headers = assign(hash.headers || {}, adapterHeaders);
+    }
+
+    const mungedOptions = mungOptionsForFetch(hash);
+
+    // Mimics the default behavior in Ember Data's `ajaxOptions`, namely to set the
+    // 'Content-Type' header to application/json if it is not a GET request and it has a body.
+    if (
+      mungedOptions.method !== 'GET' &&
+      mungedOptions.body &&
+      (mungedOptions.headers === undefined ||
+        !(
+          mungedOptions.headers['Content-Type'] ||
+          mungedOptions.headers['content-type']
+        ))
+    ) {
+      mungedOptions.headers = mungedOptions.headers || {};
+      mungedOptions.headers['Content-Type'] = 'application/json; charset=utf-8';
+    }
+
+    return mungedOptions;
+  },
+
+  /**
+   * @override
+   */
+  ajax(url: string, type: Method, options: object) {
+    const requestData = {
+      url,
+      method: type
+    };
+
+    const hash = this.ajaxOptions(url, type, options);
+
+    return (
+      this._ajaxRequest(hash)
+        // @ts-ignore
+        .catch((error, response, requestData) => {
+          throw this.ajaxError(this, response, null, requestData, error);
+        })
+        .then((response: Response) => {
+          return RSVP.hash({
+            response,
+            payload: determineBodyPromise(response, requestData)
+          });
+        })
+        .then(
+          ({
+            response,
+            payload
+          }: {
+            response: Response;
+            payload: string | object | undefined;
+          }) => {
+            if (response.ok) {
+              return this.ajaxSuccess(this, response, payload, requestData);
+            } else {
+              throw this.ajaxError(this, response, payload, requestData);
+            }
+          }
+        )
+    );
+  },
+
+  /**
+   * Overrides the `_ajaxRequest` method to use `fetch` instead of jQuery.ajax
+   * @override
+   */
+  _ajaxRequest(
+    options: Mix<RequestInit, { url: string }>
+  ): RSVP.Promise<Response> {
+    return this._fetchRequest(options.url, options);
+  },
+
+  /**
+   * A hook into where `fetch` is called.
+   * Useful if you want to override this behavior, for example to multiplex requests.
+   */
+  _fetchRequest(url: string, options: RequestInit): RSVP.Promise<Response> {
+    return fetch(url, options);
+  },
+
+  /**
+   * @override
+   */
+  ajaxSuccess(
+    adapter: any,
+    response: Response,
+    payload: Nullable<string | object>,
+    requestData: { url: string; method: string }
+  ): object | DS.AdapterError | RSVP.Promise<never> {
```

Comment:
Can we see of `RSVP.Promise<void>` works here?

Reply:
`reject(reason?: any, label?: string | undefined): RSVP.Promise<never>`
It's typed for `RSVP.rejct`

Reply:
`Promise<never>` is assignable to `Promise<void>` so you should be ok to make this change

```ts
function handlePromise(p: Promise<void>): void {

}
handlePromise(new Promise<never>(null as never))
```
https://typescript-play.js.org/#code/GYVwdgxgLglg9mABACwIZgCYBsCmAFAJzgFsYBnHACgAcAuRQk8nAHgDc4YMA+ASno5dEAbwBQogL6i0mXI1IVKYHAHcGRBa2VscBbkpBYsiVGUTbdvXkA